### ISSUE TITLE
nuget timeouts during msbuild build

### ISSUE THREAD

**Issue by kumpera** [mono/mono#9223]
How to reproduce:
1. Nuke your nuget cache: `rm -rf ~/.nuget`
1. Checkout `https://github.com/mono/msbuild xplat-master`
1. Build with `./build.sh -host mono -configuration Release -skipTests`

It will fail with this:
```
The HTTP request to ’GET https://dotnetmyget.blob.core.windows.net/artifacts/dotnet-core/nuget/v3/flatcontainer/system.security.claims/index.json' has timed out after 100000ms.
```

If you scroll up a bit you'll see: 
```
Failed to download package 'System.Runtime.Extensions.4.3.0' from 'https://api.nuget.org/v3-flatcontainer/system.runtime.extensions/4.3.0/system.runtime.extensions.4.3.0.nupkg'.
Error getting response stream (ReadAsync): ReceiveFailure Unable to read data from the transport connection: Connection reset by peer.
  Unable to read data from the transport connection: Connection reset by peer.
  Connection reset by peer
```

**Comment by marek-safar** [comment]
I can reproduce this issue (will try to get better log)

**Comment by baulig** [comment]
Questions we need to answer:

1. Is this a network / server issue?  Can we somehow confirm with the server that it is not?
2. Is nuget retrying the download (as it should) or what happens after the timeout?
3. How often does this happen?

**Comment by alexanderkyte** [comment]
1. We can confirm by seeing if older versions of mono have this problem with the current nuget server. 
2. Retries compounding with delays can lead to really long restore times, long enough that people talk about putting all their nugets into a github repo (mentioned in slack a half hour ago)
3. It seems to happen relatively often, but nuget hides this by caching packages system-wide. It particularly hurts CI and other automated builds.

**Comment by alexanderkyte** [comment]
@marek-safar if you could confirm non-reproduction with an earlier mono it might help us to reduce the number of commits that might have caused this

**Comment by kumpera** [comment]
@baulig 

1. I downloaded one of the affected nuget packages from the same network 353 times and the highest download time was 1290ms, which 96% finishing within 500ms.
2. No, it's not, it fails restoring as it probably believes that the network is down or having issues if nothing comes after 100s.
3.frequent enough. I don't have % numbers, but I repro'd it today 3 times in the course of troubleshooting a msbuild issue.

**Comment by baulig** [comment]
Fixed in #9228.
